docs/rfcs/027-pageserver-migration.md

```diff
@@ -0,0 +1,454 @@
+at scale, in several contexts:
+
+1. Responding to a pageserver node failure by migrating tenants to other pageservers
+2. Balancing load and capacity across pageservers, for example when a user expands their
+   database and they need to migrate to a pageserver with more capacity.
+3. Restarting pageservers for upgrades and maintenance
+
+Currently, a tenant may migrated by attaching to a new node,
+re-configuring endpoints to use the new node, and then later detaching from the old node. This is safe once [generation numbers](025-generation-numbers.md) are implemented, but does meet
+our seamless/fast/efficient goals:
+
+- Not fast: The new node will have to download potentially large amounts
+  of data from S3, which may take many minutes.
+- Not seamless: If we attach to a new pageserver before detaching an old one,
+  the new one might delete some objects that interrupt availability of reads on the old one.
+- Not efficient: the old pageserver will continue uploading
+  S3 content during the migration that will never be read.
+
+The user expectations for availability are:
+
+- For planned maintenance, there should be zero availability
+  gap
+- For unplanned changes (e.g. node failures), there should be
+  minimal availability gap.
+
+## Non Goals (if relevant)
+
+- We do not aim to have the pageservers fail over if the
+  control plane is unavailable.
+- On unplanned migrations (node failures), we do not aim to prevent a small, bounded window of
+  read unavailability of very recent LSNs (because postgres
+  page cache usually contains such pages, we do not expect
+  them to be read frequently from the pageserver).
+
+## Impacted components (e.g. pageserver, safekeeper, console, etc)
+
+Pageserver, control plane
+
+## Terminology
+
+- **Attachment**: a tenant is _attached_ to a pageserver if it has
+  been issued a generation number, and is running an instance of
+  the `Tenant` type, ingesting the WAL, and available to serve
+  page reads.
+- **Location**: locations are a superset of attachments. A location
+  is a combination of a tenant and a pageserver. We may _attach_ at a _location_.
+
+- **Secondary location**: a location which is not currently attached.
+- **Warm secondary location**: a location which is not currently attached, but is endeavoring to maintain a warm local cache of layers. We avoid calling this a _warm standby_ to avoid confusion with similar postgres features.
+
+## Implementation (high level)
+
+### Secondary locations
+
+To enable faster migrations, we will identify at least one _secondary location_
+for each tenant. This secondary location will keep a warm cache of layers
+for the tenant, so that if it is later attached, it can catch up with the
+latest LSN quickly.
+
+The control plane is responsible for selecting secondary locations, and
+calling into pageservers to instruct them as to which tenants they are
+to act as a secondary location for.
```

Comment:
"to instruct them as to which ...". Please come up with a verb for that. I understand "attach" as a verb is taken for the equivalent of "promote to primary".

These should also make it into the earlier glossary section.

Reply:
I think "configure" will do here: the verb will be to configure a location, and we'll have the various nouns for states.